Query refactoring: SpanOrQueryBuilder and Parser

cbuescher opened the pull request:
Moving the query building functionality from the parser to the builders
new doToQuery() method analogous to other recent query refactorings.

Relates to #10217

MaineC commented on core/src/main/java/org/elasticsearch/index/query/SpanOrQueryBuilder.java:
```diff
@@ -19,24 +19,38 @@
+import java.util.List;
+import java.util.Objects;
 
 public class SpanOrQueryBuilder extends AbstractQueryBuilder<SpanOrQueryBuilder> implements SpanQueryBuilder<SpanOrQueryBuilder> {
```
Nitpick: With this turning into a public API - should we have some class level doc here?

MaineC commented on core/src/main/java/org/elasticsearch/index/query/SpanOrQueryParser.java:
```diff
@@ -34,7 +31,7 @@
 /**
  *
  */
```
Either fill with real information or remove the empty comment.

MaineC commented on core/src/main/java/org/elasticsearch/index/query/SpanOrQueryParser.java:
```diff
@@ -34,7 +31,7 @@
-public class SpanOrQueryParser extends BaseQueryParserTemp {
+public class SpanOrQueryParser extends BaseQueryParser {
 
     @Inject
```
Shouldn't be needed for a constructor w/o arguments.

MaineC commented on core/src/test/java/org/elasticsearch/index/query/SpanOrQueryBuilderTest.java:
```diff
@@ -0,0 +1,82 @@
+        SpanOrQueryBuilder queryBuilder = new SpanOrQueryBuilder();
+        int clauses = randomIntBetween(1, 6);
+        // we use one random SpanTermQueryBuilder to determine same field name for subsequent clauses
+        String fieldName = new SpanTermQueryBuilderTest().createTestQueryBuilder().fieldName();
```
Is this something that could be moved to a more general utility class for testing span queries?

W/o having looked at the implementation: Or at least something that could be moved to a static helper method of the test class itself?

cbuescher replied:
Not quiet sure which part you think is general enough to factor our here. I briefly tried having a general Span-Query producing utility method in RandomQueryBuilder but the problem here is that ony SpanTermQueryBuilder is a leaf query and for the others we would have to deal with the depth of nested subtrees. Just getting a list of span queries with the same fieldname is only needed here, the other span queries come either with one or two subclauses only.

cbuescher replied:
I my last commit (after rebase) I added a utility method to SpanTermQueryBuilderTest that outputs an array of random SpanTermQueryBuilders with same fieldname and configurable size. Let me know what you think, I was able to reuse that in the other tests as well but maybe it's confusing? Let me know what you think.

MaineC commented:
Left some minor comments.

cbuescher commented:
@MaineC thanks for the review, just pushed new commit addressing your comments.

MaineC commented:
LGTM - I'm sure @javanna will will find a few more areas for improvment

javanna commented on core/src/main/java/org/elasticsearch/index/query/SpanOrQueryBuilder.java:
```diff
@@ -53,6 +70,55 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
+        SpanQuery[] spanQueries = new SpanQuery[clauses.size()];
+        for (int i = 0; i < clauses.size(); i++) {
+            Query query = clauses.get(i).toQuery(parseContext);
+            assert query instanceof SpanQuery;
```
we might have the same problem with LateParsingQuery here too right? then this should become an if instead? I guess we might have the same somewhere else too?

cbuescher replied:
Not really, all query clauses here need to be SpanQueryBuilders, the problem with LateParsingQuery was related to MultiTermQueryBuilders (which RangeQuery is a subtype of). We are creating SpanOrQuery here which takes list of SpanQueries as an argument, so all good I think.

javanna replied:
ok sorry for the confusion, here and in the other PR where I commented...

javanna commented:
left one small comment

cbuescher commented:
Addressed comments and removed check for empty clauses in doXContent since this will either be caught by parser or in validate. Left one question regarding tests to get another opinion from @MaineC.

MaineC commented:
After checking the modifications in person with @cbuescher this morning: Looks all good to me.
